### ISSUE TITLE
Displaying many algorithms

### ISSUE THREAD

**Issue by brockho** [numbbo/coco#1370]
After a discussion with @nikohansen, it turns out that probably the best API from python to do this is the following:

We will have a mandatory argument of `main` with the algorithms to display (in the foreground) plus an optional argument with the algorithms to display in the background. 

This has the advantage that it can be combined with the current functionality to get the algorithms that contribute to the artificial best in a first step. (This code part has to be adapted before to output these two sets of algorithms, i.e. background and foreground algorithms.)

**Comment by brockho** [comment]
In order to make the postprocessing quicker, we might want to consider pre-preparing all plots with the background and then only display the highlighted algorithms on top of this. The following link gives some idea how to actually do that: http://joergdietrich.github.io/matplotlib-overplot.html

**Comment by nikohansen** [comment]
Apparently, `imread` and `imshow` are the functions to make this work, see [here](https://stackoverflow.com/questions/15160123/adding-a-background-image-to-a-plot-with-known-corner-coordinates#15160269)

**Comment by dtusar** [comment]
Now you can specify which algorithms are in background with setting accordingly the `background_algorithms` in `genericsettings.py`.

**Comment by nikohansen** [comment]
It would be useful to have an out-commented example or a comment how algorithms can be specified (I assume it's a folder or zip file name?).

**Comment by dtusar** [comment]
It's the same path as the input parameter(s) you use for retrieving the data.

**Comment by nikohansen** [comment]
It should be documented anyways (done).

Adding example code of a use case in this issue may also be helpful for us.

**Comment by brockho** [comment]
An example will be indeed needed. I tried with 
```
background_algorithms =['ACOR_liao_noiseless.tgz', 'ALPS_hornby_noiseless.tgz', 'AMALGAM_bosman_noiseless.tgz',
 'BAYEDA_gallagher_noiseless.tgz', 'BFGS_ros_noiseless.tgz', 'BIPOP-aCMA-STEP_loshchilov_noiseless.tgz',
 'BIPOP-CMA-ES_hansen_noiseless.tgz', 'BIPOP-saACM-k_loshchilov_noiseless.tgz', 'BIPOPaCMA_loshchilov_noiseless.tgz']
```
with all specified files in the current folder, but postprocessing 
```
> python -m cocopp -o ppdataBackgroundTest-2 --omit-single BIPOP-CMA-ES_hansen_noiseless.tgz RANDOMSEARCH_auger_noiseless.tgz SMAC-BBOB_hutter_noiseless.tgz
```
only gives me two colored and one gray graph in each ECDF figure:

![pprldmany_10d_noiselessall](https://user-images.githubusercontent.com/13706996/27285703-d4057232-54fd-11e7-85bf-42da305012cb.png)

This must be considered as a bug.

Note that the three data sets for BIPOP-CMA-ES, RANDOMSEARCH and SMAC-BBOB have been extracted already before the above call to the postprocessing, the others (for the background) are still kept as `.tgz` files and are not extracted at all. If I now remove the BIPOP-CMA-ES entry from the `background_algorithms`, I get the three algorithms in the foreground, but nothing in the background:

![pprldmany_10d_noiselessall-2ndtry](https://user-images.githubusercontent.com/13706996/27285913-94649b8e-54fe-11e7-9202-c30a0969e0d0.png)

Even when I also extract the algorithms in the `background_algorithms` and specify all extracted folders within `background_algorithms`, the ECDF figure does not show the background algorithms! Taking out the to-be-displayed algorithms from `background_algorithms` also does not work. All this holds independently of whether the extracted folders start with `.extracted` or not.

To summarize, we need as a first steps:
* that data files, specified in `background_algorithms` are extracted in the same way normal input arguments to the postprocessing are extracted
* that algorithms, specified in the `background_algorithms` are actually taken into account (in particular if their data is available in the folder where the call to `cocopp` is made)
* that algorithms that appear both in the set of algorithms to display and in the `background_algorithms` will be displayed in the highlighted color and not only in the background style (i.e. give precedence of to-be-plotted algorithms over the ones specified in the `background_algorithms`

**Comment by nikohansen** [comment]
Isn't there a problem in that `cocopp.genericsettings` is not available when called from the command line? We most likely don't want to edit the file of the installed `cocopp` python module. Currently I only see any reasonable use case scenario from the Python shell.

**Comment by brockho** [comment]
The above issue seems be solved with the latest commits. However, there are still things to improve :-)

For example should the labels in the following plot not be displayed on top of each other but in a readable manner.

![image](https://user-images.githubusercontent.com/13706996/27389057-9bff58fc-569d-11e7-82c6-b3ac17425352.png)

Furthermore, I am not sure whether all background algorithms shouldn't get the same line styles (solid)?

**Comment by nikohansen** [comment]
Agreed, the background algorithms should have the same line style. I am right that we can display different groups of background algorithms with different colours and/or line styles?

**Comment by dtusar** [comment]
> I am right that we can display different groups of background algorithms with different colours and/or line styles?

Not yet.

**Comment by dtusar** [comment]
Now we can define different groups of background algorithms, the definition has slightly changed (following the proposal above). There is an example in `genericsettings.py` how to define background algorithms. 

The labels are also fixed.

**Comment by brockho** [comment]
Seems to work indeed. Many thanks @dtusar for your work! 

I added an a bit more detailed example on how to use the new functionality within `genericsettings.py` to not forget the colors, we had before. I hope it is okay to close the issue now.

**Comment by brockho** [comment]
It seems that the foreground algorithms are not anymore plotted last in the current version of the postprocessing:

![image](https://user-images.githubusercontent.com/13706996/29805767-fcd07688-8c8b-11e7-86f3-d1bfd130b758.png)

This makes the plots again unreadable. Hence, the reopening of the issue.

**Comment by brockho** [comment]
Since this issue does not show up in the current development branch at the moment, I'll close the issue again.

**Comment by nikohansen** [comment]
How do I make all 2009 algorithms show up in the background, as it was at some point showing up by default?

**Comment by brockho** [comment]
I cannot remember that this was default at any time (only with the very old `.pickle` files and the old ECDF plots). But for me, earlier today
```
cocopp.genericsettings.background = {('#bbbbbb', '-'): cocopp.bbob.get('2009')}
```
did what you want. I am not sure, however, if this works for a single algorithm (in the sense that it might not have any effect).

**Comment by nikohansen** [comment]
WOW, that looks kind-of cool.

**Comment by brockho** [comment]
To add: the first entry `#XXXXXX` indicates the color (in hexadecimal RGB, e.g. `#ff0000` is red) and the second entry of the tupel describes the line style (e.g. `-` for solid lines, `--` for dashed ones, etc.).

**Comment by dtusar** [comment]
> I am not sure, however, if this works for a single algorithm (in the sense that it might not have any effect).

Yes, this should also work for a single algorithm.

**Comment by nikohansen** [comment]
BTW, if we set, for the time being
```python
cocopp.genericsettings.background = {('#bbbbbb', '-'): cocopp.bbob.get_all('2009', remote=False)}
```
There will be no download overhead.
